src/main/scala/org/hammerlab/guacamole/commands/StructuralVariantCaller.scala

```diff
@@ -0,0 +1,133 @@
+  }
+
+  case class Location(
+    contig: String,
+    position: Long) {
+  }
+
+  case class GenomeRange(
+    contig: String,
+    start: Long,
+    stop: Long
+  )
+
+  object Caller extends SparkCommand[Arguments] {
+    override val name = "structural-variant"
+    override val description = "Find structural variants, i.e. large insertions or deletions, inversions and translocations."
+
+    def isFirstInPair(read: MappedRead) = {
+      read.matePropertiesOpt match {
+        case Some(mateProps) => mateProps.isFirstInPair
+        case None            => false
+      }
+    }
+
+    // The sign of inferredInsertSize follows the orientation of the read. This returns
+    // an insert size which is positive in the common case, regardless of the orientation.
+    def orientedInsertSize(r: MappedRead): Option[Int] = {
+      r.matePropertiesOpt.flatMap(_.inferredInsertSize.map(x => x * (if (r.isPositiveStrand) { 1 } else { -1 })))
+    }
+
+    // Coalesce sequences of values separated by blockSize into single ranges.
+    // The sequence of values must be sorted.
+    def coalesceAdjacent(xs: IndexedSeq[Long], blockSize: Long): Iterator[(Long, Long)] = {
+      val rangeStarts = ArrayBuffer(0)
+      for (i <- 1 until xs.length) {
+        if (xs(i - 1) + blockSize != xs(i)) {
+          rangeStarts += i
+        }
+      }
+      rangeStarts += xs.length
+
+      rangeStarts.sliding(2).map(items => {
+        val startIdx = items(0)
+        val nextStartIdx = items(1)
+        (xs(startIdx), xs(nextStartIdx - 1))
+      })
+    }
+
+    // Returns a sequence of all blocks which overlap the read, its mate or the insert between them.
+    def matedReadBlocks(r: MappedRead): Seq[Long] = {
+      val mateStart = r.matePropertiesOpt.map(_.mateStart.getOrElse(r.start)).getOrElse(r.start)
+      val start = List(r.start, r.end, mateStart).min
+      val roundedStart = (1.0 * start / BLOCK_SIZE).toLong * BLOCK_SIZE
+      val insertSize = orientedInsertSize(r).getOrElse(0)
+
+      (roundedStart to (roundedStart + insertSize) by BLOCK_SIZE)
+    }
+
+    override def run(args: Arguments, sc: SparkContext): Unit = {
+      val readSet = Common.loadReadsFromArguments(args, sc, Read.InputFilters(nonDuplicate = true))
+      val pairedMappedReads = readSet.mappedReads.filter(_.isPaired)
+      val firstInPair = pairedMappedReads.filter(isFirstInPair)
+
+      // Pare down to mate pairs which aren't highly displaced & aren't inverted.
+      val readsInRange = firstInPair.filter(r => {
+        val insertSize = orientedInsertSize(r).getOrElse(0)
+        (insertSize > 0) && (insertSize < MAX_INSERT_SIZE)
+      })
+      readsInRange.persist()
+
+      val insertSizes = readsInRange.flatMap(orientedInsertSize)
+      val insertStats = insertSizes.stats()
+      println("Stats on inferredInsertSize: " + insertStats)
+
+      // Create a map from (block) -> [insert sizes]
+      // TODO: take greater advantage of data locality than .groupByKey() does?
```

Comment:
Yea, I think either one will work; if I have it right, `combineByKey` is more general, but for this case all you need is `aggregateByKey`.

You are going from an `RDD[(K, V)]` to an `RDD[(K, U)]`. You have:
* a `zeroValue: U`,
* a `seqOp: (U, V) => U`, and
* a `combOp: (U, U) => U`,

which are the necessities for the `aggregateByKey` API.

`combineByKey` generalizes further by substituting a `createCombiner: V => U` instead of `zeroValue: U`.

`K == Location`, `V == Long`, `U == StatsCounter`, `zeroValue == new StatCounter()`, and `seqOp` and `combOp` are overloaded versions of `StatsCounter.merge` that take a `Double` and a `StatsCounter`, resp.